pre-aio-node-with-zuul-ref.sh

```diff
@@ -0,0 +1,157 @@
+NOVA_CONF_PATH=/etc/nova
+SAMPLE_IMAGE=cirros-0.3.5-x86_64-disk
+VM_NAME="testVM"
+# Set up internal variables
+_SSH_OPTIONS="\
+    -o BatchMode=yes \
+    -o StrictHostKeyChecking=no \
+    -o UserKnownHostsFile=/dev/null \
+    -i /opt/stack/.ssh/id_rsa"
+
+server_info=$(grep "VNCSERVER_PROXYCLIENT_ADDRESS" local.conf)
+server_info=${server_info#*\"}
+server_info=${server_info%\"}
+echo "servier ip:"
+echo $server_info
+
+function on_xenserver() {
+    ssh $_SSH_OPTIONS "root@$server_info" bash -s --
+}
+
+cd $DEVSTACK_TOOL/
+./clean.sh
+git checkout master
+git pull --ff origin master
+
+sed -i '/[[local|localrc]]/a DEST=/opt/stack/openstack' ${DEVSTACK_TOOL}/local.conf
+
+IFS='^' read -r -a ZUUL_PROJ_CHANGE <<< "$ZUUL_CHANGES"
+related_proj=()
+for proj in ${ZUUL_PROJ_CHANGE[@]}
+do
+    proj=${proj%%:*}
+    proj=${proj##*/}
+    related_proj+=($proj)
+done
+unique_proj_list=$(echo "${related_proj[@]}" | tr ' ' '\n' | nl | sort -u -k2 | sort -n | cut -f2-)
+
+echo ${DEST}
+cd ${DEST}
+
+for dir in ${DEST}/*
+do
+    if [ -d "${dir}" ] ; then
+        echo "check ${dir}"
+        cd ${dir}
+        if [ -d .git ]; then
+            echo "update repository ${dir}"
+            git checkout master
+            git pull --ff origin master
+
+            pkg_flag=${dir##*/}
+            echo $pkg_flag
+            if [[ " ${unique_proj_list[@]} " =~ " ${pkg_flag} " ]]; then
+                echo "merge repository ${dir}"
+                git branch -D $TMP_MERGE_BRANCH
+                git checkout master -B $TMP_MERGE_BRANCH
+                git fetch ${FETCH_URL}/openstack/${pkg_flag} $refspec
+                git merge FETCH_HEAD -m "temp branch for vgpu patches merge"
+            fi
+        fi;
+    fi
+done
+
+first_vgpu_type=$(on_xenserver << END_OF_REQ_VGPU_TYPE
+vgpu_type_list=\$(xe vgpu-type-list --minimal)
+vgpu_type_list=\${vgpu_type_list//,/ }
+for id in \$vgpu_type_list
+do
+    name=\$(xe vgpu-type-param-get uuid=\$id param-name="model-name")
+    if [ "\$name" != "passthrough" ]
+    then
+        echo \$name
+        exit
+    fi
+done
+END_OF_REQ_VGPU_TYPE
+)
+
+cd $DEVSTACK_TOOL/
+./stack.sh
+echo "[devices]" >> ${NOVA_CONF_PATH}/nova.conf
+echo "enabled_vgpu_types = $first_vgpu_type" >> ${NOVA_CONF_PATH}/nova.conf
+sudo systemctl restart devstack@n-cpu.service
+
+
+cd $TEMPEST_DIR
+# TODO: run vgpu tempest
+
+source $DEVSTACK_TOOL/openrc admin
+nova flavor-key 1 set resources:VGPU=1
+image_id=$(glance image-list | grep "$SAMPLE_IMAGE" | awk '{print $2}')
+glance image-update  --property vm_mode=hvm $image_id
+
+source $DEVSTACK_TOOL/openrc demo demo
+prv_net=$(openstack network list | grep "private" | awk '{print $2}')
+
+nova boot --image $SAMPLE_IMAGE --flavor 1 --nic net-id=$prv_net $VM_NAME
+
+nova_vm_id=$(nova show $VM_NAME | grep -w 'id' | awk '{print $4}')
+
+i=0
+while :
+do
+    sleep 5
+    i=`expr $i + 1`
+    vm_state=$(nova list  --name $VM_NAME --fields status | grep ACTIVE)
```

Comment:
GET the status, if it's FAILED, we can stop earlier. With reasonable failed reason printed.

Reply:
OK